### ISSUE TITLE
block gzip files cannot be read in glow v1.0.0

### ISSUE THREAD

**Issue by williambrandler** [projectglow/glow#354]
When reading block gzipped vcf or gff in glow v1.0.0, I get

`import glow
spark = glow.register(spark)
display(spark.read.format("vcf").load('dbfs:/databricks-datasets/genomics/jg-sample/HG00097.g.vcf.bgz'))`

`org.apache.spark.SparkException: Job aborted due to stage failure: Task 255 in stage 0.0 failed 4 times, most recent failure: Lost task 255.3 in stage 0.0 (TID 258) (10.126.235.231 executor 10): htsjdk.tribble.TribbleException$InvalidHeader: Your input file has a malformed header: We never saw the required CHROM header line (starting with one #) for the input VCF file`

Same issue with different error for gff,

`gff_path = "/databricks-datasets/genomics/gffs/GCF_000001405.39_GRCh38.p13_genomic.gff.bgz"
original_gff_df = spark.read \
  .format("gff") \
  .load(gff_path)`

`
org.apache.spark.SparkException: Job aborted due to stage failure: Task 10 in stage 15.0 failed 4 times, most recent failure: Lost task 10.3 in stage 15.0 (TID 70, 10.148.200.234, executor 2): java.lang.RuntimeException: Duplicate map key H�� was found, please check the input data. If you want to remove the duplicated keys, you can set spark.sql.mapKeyDedupPolicy to LAST_WIN so that the key inserted at last takes precedence.`

Works fine for files with `.gz` extension

**Comment by williambrandler** [comment]
`Py4JJavaError: An error occurred while calling o265.load.
: org.apache.spark.SparkException: Job aborted due to stage failure: Task 255 in stage 0.0 failed 4 times, most recent failure: Lost task 255.3 in stage 0.0 (TID 258) (10.126.235.231 executor 10): htsjdk.tribble.TribbleException$InvalidHeader: Your input file has a malformed header: We never saw the required CHROM header line (starting with one #) for the input VCF file
	at htsjdk.variant.vcf.VCFCodec.readActualHeader(VCFCodec.java:115)
	at io.projectglow.vcf.VCFFileFormat$.$anonfun$createVCFCodec$2(VCFFileFormat.scala:264)
	at io.projectglow.common.WithUtils$.withCloseable(WithUtils.scala:46)
	at io.projectglow.vcf.VCFFileFormat$.$anonfun$createVCFCodec$1(VCFFileFormat.scala:261)
	at io.projectglow.common.WithUtils$.withCloseable(WithUtils.scala:46)
	at io.projectglow.vcf.VCFFileFormat$.createVCFCodec(VCFFileFormat.scala:253)
	at io.projectglow.vcf.VCFHeaderUtils$.$anonfun$createHeaderRDD$1(VCFHeaderUtils.scala:98)
	at scala.collection.Iterator$$anon$11.nextCur(Iterator.scala:484)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:490)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:489)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at org.apache.spark.util.collection.ExternalSorter.insertAll(ExternalSorter.scala:192)
	at org.apache.spark.shuffle.sort.SortShuffleWriter.write(SortShuffleWriter.scala:62)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.doRunTask(Task.scala:150)
	at org.apache.spark.scheduler.Task.run(Task.scala:119)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$10(Executor.scala:732)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1643)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:735)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2711)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2658)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2652)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2652)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1255)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1255)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1255)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2919)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2860)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2848)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:1029)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2439)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2460)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2479)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2504)
	at org.apache.spark.rdd.RDD.$anonfun$collect$1(RDD.scala:1035)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:125)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
	at org.apache.spark.rdd.RDD.withScope(RDD.scala:419)
	at org.apache.spark.rdd.RDD.collect(RDD.scala:1034)
	at io.projectglow.vcf.VCFHeaderUtils$.getUniqueHeaderLines(VCFHeaderUtils.scala:130)
	at io.projectglow.vcf.VCFHeaderUtils$.readHeaderLines(VCFHeaderUtils.scala:170)
	at io.projectglow.vcf.SchemaDelegate$.io$projectglow$vcf$SchemaDelegate$$readHeaders(VCFFileFormat.scala:441)
	at io.projectglow.vcf.SchemaDelegate$FlattenedInfoDelegate.schema(VCFFileFormat.scala:472)
	at io.projectglow.vcf.VCFFileFormat.inferSchema(VCFFileFormat.scala:82)
	at org.apache.spark.sql.execution.datasources.DataSource.$anonfun$getOrInferFileFormatSchema$11(DataSource.scala:230)
	at scala.Option.orElse(Option.scala:447)
	at org.apache.spark.sql.execution.datasources.DataSource.getOrInferFileFormatSchema(DataSource.scala:223)
	at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:460)
	at org.apache.spark.sql.DataFrameReader.loadV1Source(DataFrameReader.scala:408)
	at org.apache.spark.sql.DataFrameReader.$anonfun$load$3(DataFrameReader.scala:390)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:390)
	at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:277)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:380)
	at py4j.Gateway.invoke(Gateway.java:295)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:251)
	at java.lang.Thread.run(Thread.java:748)
Caused by: htsjdk.tribble.TribbleException$InvalidHeader: Your input file has a malformed header: We never saw the required CHROM header line (starting with one #) for the input VCF file
	at htsjdk.variant.vcf.VCFCodec.readActualHeader(VCFCodec.java:115)
	at io.projectglow.vcf.VCFFileFormat$.$anonfun$createVCFCodec$2(VCFFileFormat.scala:264)
	at io.projectglow.common.WithUtils$.withCloseable(WithUtils.scala:46)
	at io.projectglow.vcf.VCFFileFormat$.$anonfun$createVCFCodec$1(VCFFileFormat.scala:261)
	at io.projectglow.common.WithUtils$.withCloseable(WithUtils.scala:46)
	at io.projectglow.vcf.VCFFileFormat$.createVCFCodec(VCFFileFormat.scala:253)
	at io.projectglow.vcf.VCFHeaderUtils$.$anonfun$createHeaderRDD$1(VCFHeaderUtils.scala:98)
	at scala.collection.Iterator$$anon$11.nextCur(Iterator.scala:484)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:490)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:489)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at org.apache.spark.util.collection.ExternalSorter.insertAll(ExternalSorter.scala:192)
	at org.apache.spark.shuffle.sort.SortShuffleWriter.write(SortShuffleWriter.scala:62)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
	at org.apache.spark.scheduler.Task.doRunTask(Task.scala:150)
	at org.apache.spark.scheduler.Task.run(Task.scala:119)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$10(Executor.scala:732)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1643)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:735)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more`

**Comment by karenfeng** [comment]
I encounter the same issue if the following Spark configuration is not set:
```
spark.hadoop.io.compression.codecs=io.projectglow.sql.util.BGZFCodec
```
Unfortunately, this is a little hard to find the [documentation](https://glow.readthedocs.io/en/latest/getting-started.html).

I'm not sure if #326 attempts to automatically register the codec. @henrydavidge, can you clarify?

**Comment by karenfeng** [comment]
Is the BGZF codec set? The [docs](https://glow.readthedocs.io/en/latest/getting-started.html) suggest the following:
```
--conf spark.hadoop.io.compression.codecs=io.projectglow.sql.util.BGZFCodec
```

**Comment by williambrandler** [comment]
ah thanks Karen,
Do you think it is worth adding the spark conf in the starter notebooks?
I plan to create a PR around updating notebooks, wondering whether to add this?

**Comment by williambrandler** [comment]
Tested out, setting the spark conf when setting up the cluster worked!
Thanks @karenfeng!

**Comment by henrydavidge** [comment]
Hm, interesting. I'm surprised that the automatic registration isn't working. Let me take a look.

**Comment by williambrandler** [comment]
this has been resolved in glow v1.0.1

**Comment by ptn24** [comment]
+1
